### PR TITLE
Adds a data collection utility and improves support for loading in MFEMSidreDataCollection

### PR DESCRIPTION
**Pull request by kennyweiss**
# Summary

- This PR is a feature, but also includes some bugfixes
- It adds a new tool for generating distributed blueprint-compliant (high order) meshes over a Cartesian domain or from an mfem mesh: `data_collection_util`. The latter is active when Axom is configured with `mfem` and `AXOM_ENABLE_MFEM_SIDRE_DATACOLLECTION`
 
- This PR also improves support for loading data collections in `MFEMSidreDataCollection`. 
- Specifically:
    - It avoids appending the `.root` suffix to an input root file when that suffix is already present
    - It wires up the high order nodal grid function during load (when available)
    - It improves error messages when attempting to load using a different number of ranks than were used to save the file

- This PR also add a new string utility function: `removeSuffix()`
- I noticed that the string utilities did not have unit tests, so I added some
- In doing so, there were a few undocumented edge-case behaviors. I'd like the reviewers to sign off on the changes I made (which I'll point out via comments)

#### TODO:

- [x] Update RELEASE-NOTES

### CODE REVIEW

File: src/axom/core/utilities/StringUtilities.cpp [hunk 1]
```diff
@@ -48,15 +48,20 @@ std::vector<std::string> splitLastNTokens(const std::string& input,
   auto last_pos = std::string::npos;
   auto pos = input.find_last_of(delim, last_pos - 1);
 
-  while((pos != std::string::npos) && (result.size() < n - 1))
+  if(n > 0 && !input.empty())
```
Comment (kennyweiss) [inline src/axom/core/utilities/StringUtilities.cpp:51]: @joshessman-llnl -- I think you added this function. Can you please confirm the new edge-case conditions.

Essentially, if the input is empty or the user asks for 0 tokens, the function now returns an empty vector (instead of a vector with an empty string, and a vector with all tokens, respectively)

File: src/axom/core/tests/utils_utilities.hpp [hunk 2]
```diff
@@ -91,16 +91,9 @@ TEST(utils_Utilities, random_real_with_seed)
   constexpr double a = -5.0;
   constexpr double b = 5.0;
 
-  const double expected_reals[5] = {-1.5112829544380526,
-                                    -2.3311429024686219,
-                                    -3.6335370551231403,
-                                    -4.714431326610093,
-                                    3.6893326916732878};
```
Comment (kennyweiss) [inline src/axom/core/tests/utils_utilities.hpp:98]: Two comments here:
1. As far as I can tell, we cannot assume that different compilers will provide the same set of values for a uniform distribution when given the same seed.  See, e.g. https://stackoverflow.com/questions/28607889/how-to-generate-the-same-random-number-sequence-over-multiple-types-of-compilers
2. For some reason, the two tests in this filer were failing when I ran the `./core_serial_tests`, but not when I ran `ctest`?

File: src/axom/sidre/core/MFEMSidreDataCollection.cpp [hunk 3]
```diff
@@ -855,13 +855,23 @@ void MFEMSidreDataCollection::Load(const std::string& path,
...
+
   #if defined(AXOM_USE_MPI) && defined(MFEM_USE_MPI)
   if(m_comm != MPI_COMM_NULL)
   {
-    IOManager reader(m_comm);
     // The conduit abstraction appears to automatically handle the ".root"
     // suffix, but the IOManager does not, so it gets added here
-    reader.read(m_bp_grp->getDataStore()->getRoot(), path + ".root");
+    using axom::utilities::string::endsWith;
+    std::string suffixedPath = endsWith(path, ".root") ? path : path + ".root";
```
Comment (kennyweiss) [inline src/axom/sidre/core/MFEMSidreDataCollection.cpp:867]: DC bugfix 1: Don't append `.root` if it's already there.

File: src/axom/sidre/core/MFEMSidreDataCollection.cpp [hunk 4]
```diff
@@ -855,13 +855,23 @@ void MFEMSidreDataCollection::Load(const std::string& path,
...
+
+    IOManager reader(m_comm);
+    reader.read(m_bp_grp->getDataStore()->getRoot(), suffixedPath);
+
+    // Get some data in support of error checks below
+    numInputRanks = reader.getNumGroupsFromRoot(suffixedPath);
+    MPI_Comm_size(m_comm, &numCommRanks);
```
Comment (kennyweiss) [inline src/axom/sidre/core/MFEMSidreDataCollection.cpp:874]: DC bugfix/enhancement 2: Improved logging (a few lines down), if loading on a different number of ranks than were used for save.
Note @nselliott: this required making a private function from `IOManager` public (`getNumGroupsFromRoot()`)

File: src/axom/core/utilities/StringUtilities.cpp (hunk 1 above)
Comment (joshessman-llnl) [reply src/axom/core/utilities/StringUtilities.cpp:51]: I think this is reasonable - it looks like the `result.size() < n - 1` was bogus anyways if `n == 0` since `n` is of unsigned type.

File: src/axom/sidre/core/MFEMSidreDataCollection.cpp [hunk 5]
```diff
@@ -873,21 +883,36 @@ void MFEMSidreDataCollection::Load(const std::string& path,
...
+  const std::string gfPath = "topologies/mesh/grid_function";
+  if(GetBPGroup()->hasView(gfPath))
+  {
+    const std::string nodalGFName = GetBPGroup()->getView(gfPath)->getString();
+    if(this->HasField(nodalGFName))
+    {
+      this->SetMeshNodesName(nodalGFName);
+      mesh->NewNodes(*this->GetField(nodalGFName), false);
+    }
+  }
```
Comment (kennyweiss) [inline src/axom/sidre/core/MFEMSidreDataCollection.cpp:915]: DC bugfix 3:  We need to account for the nodal grid function when loading an `MFEMSidreDataCollection`.

File: src/axom/core/utilities/StringUtilities.cpp (hunk 1 above)
Comment (kennyweiss) [reply src/axom/core/utilities/StringUtilities.cpp:51]: Thanks -- it was, but had the effect of making `N` irrelevant, and returned all tokens

Comment (kennyweiss) [reply src/axom/core/utilities/StringUtilities.cpp:51]: As an aside, the API/signature for this function is different from the one for `split()` above.
This one returns a vector, the other has a vector as an `out` parameter.

In the future, we might want to make these consistent.

File: src/axom/core/tests/utils_stringUtilities.hpp [hunk 6]
```diff
@@ -0,0 +1,307 @@
...
+    std::string testString = "FoO.bAr";
+    std::string expString = "foo.bar";
+    axom::utilities::string::toLower(testString);
+    EXPECT_EQ(expString, testString);
+  }
+}
+
+TEST(utils_stringUtilities, toUpper)
+{
+  // already upper
```
Comment (white238) [inline src/axom/core/tests/utils_stringUtilities.hpp:130]: I think these comments are wrong

File: src/tools/CMakeLists.txt [hunk 7]
```diff
@@ -68,6 +68,34 @@ if(AXOM_ENABLE_QUEST)
             )
 endif()
 
+#------------------------------------------------------------------------------
+# data_collection_util is a utility that generates blueprint-conforming sidre
+# output files from an input mfem mesh
+#------------------------------------------------------------------------------
+if( AXOM_ENABLE_SIDRE AND HDF5_FOUND AND MFEM_FOUND AND AXOM_ENABLE_MFEM_SIDRE_DATACOLLECTION )
```
Comment (white238) [inline src/tools/CMakeLists.txt:75]: Does HDF5_FOUND need to be checked here?

File: src/tools/data_collection_util.cpp [hunk 8]
```diff
@@ -0,0 +1,620 @@
...
+  void fixBoxParams()
+  {
+    const bool dimProvided = boxDim > 0;
+    const bool rangeProvided = !(boxMins.empty() && boxMaxs.empty());
+    const bool resProvided = !boxResolution.empty();
+
+    // First check: we need at least one of: dimension, mins/maxs, resolution
+    if(!(dimProvided || rangeProvided || resProvided))
+    {
+      throw "Box mesh must have at least one of: dimension, mins and maxs or "
```
Comment (white238) [inline src/tools/data_collection_util.cpp:77]: Is there a reason you don't use SLIC here?  If so can you add a comment?

File: src/tools/data_collection_util.cpp [hunk 9]
```diff
@@ -0,0 +1,620 @@
...
+    int szMins = boxMins.size();
+    int szMaxs = boxMaxs.size();
+    int szRes = boxResolution.size();
+
+    // Error checking on provided dimension
+    if(dimProvided)
+    {
+      if(boxDim < 1 || boxDim > 3)
+      {
+        throw "Dimension cannot exceed 3";
```
Comment (white238) [inline src/tools/data_collection_util.cpp:90]: Error is confusing if you end up with boxDim == 0

**Comment by white238** [comment]
Is there a possible test you could add for the new tool?  If not could you at least add something to make sure that it doesn't crash on a basic input?

File: src/axom/sidre/core/MFEMSidreDataCollection.cpp [hunk 10]
```diff
@@ -873,21 +883,36 @@ void MFEMSidreDataCollection::Load(const std::string& path,
...
+                  "dataset was created using {} ranks",
+                  numCommRanks,
+                  numInputRanks));
 
     UpdateStateFromDS();
     UpdateMeshAndFieldsFromDS();
   }
+
+  // Set the mesh nodal grid function when present
+  const std::string gfPath = "topologies/mesh/grid_function";
```
Comment (joshessman-llnl) [inline src/axom/sidre/core/MFEMSidreDataCollection.cpp:906]: Suggestion: I haven't gone through and replaced it everywhere, but it might be good to switch out the hardcoded "mesh" here for `MFEMSidreDataCollection::s_mesh_topology_name`.

File: src/tools/data_collection_util.cpp [hunk 11]
```diff
@@ -0,0 +1,620 @@
...
+      }
+    }
+  }
+
+public:
+  Input() = default;
+
+  bool isVerbose() const { return m_verboseOutput; }
+
+  void parse(int argc, char** argv, CLI::App& app)
```
Comment (joshessman-llnl) [inline src/tools/data_collection_util.cpp:149]: This might be a Maslow's hammer-type situation but to me this looks like a lot of command-line options; had you given any thought to using Inlet here?

**Review by joshessman-llnl — APPROVED** [review]
Thanks for catching/fixing these SidreDC bugs!

File: src/axom/core/tests/utils_stringUtilities.hpp (hunk 6 above)
Comment (kennyweiss) [reply src/axom/core/tests/utils_stringUtilities.hpp:130]: Thanks -- copy-paste error

File: src/tools/CMakeLists.txt (hunk 7 above)
Comment (kennyweiss) [reply src/tools/CMakeLists.txt:75]: It's necessary for writing out the files. 

- [ ] Make HDF5 optional and change the protocol when HDF5's not available

File: src/tools/data_collection_util.cpp (hunk 8 above)
Comment (kennyweiss) [reply src/tools/data_collection_util.cpp:77]: I'm using exceptions since the CLI parsing is in a try/catch loop

TODO:
- [ ] Add a comment about exceptions
- [ ] Confirm that this is caught properly

File: src/tools/data_collection_util.cpp (hunk 9 above)
Comment (kennyweiss) [reply src/tools/data_collection_util.cpp:90]: Agreed.

TODO:
- [ ] Fix this

**Review by agcapps — APPROVED** [review]
Thanks @kennyweiss

File: src/tools/CMakeLists.txt (hunk 7 above)
Comment (kennyweiss) [reply src/tools/CMakeLists.txt:75]: I wasn't actually setting the protocol -- just using the default.

It looks like the MFEMSIdreDataCollection should be updated to deal with cases where HDF5s' not enabled. I'll create a ticket for that so we can handle it later.

Comment (kennyweiss) [reply src/tools/CMakeLists.txt:75]: Update: https://github.com/LLNL/axom/issues/623

File: src/tools/data_collection_util.cpp (hunk 8 above)
Comment (kennyweiss) [reply src/tools/data_collection_util.cpp:77]: Thanks for the comment.

Turns out I needed to throw `CLI::ValidationError` instead of just a string.
I also updated the doxy comment to indicate that the function could throw that exception.

File: src/axom/core/utilities/StringUtilities.hpp [hunk 12]
```diff
@@ -63,6 +63,21 @@ inline bool startsWith(const std::string& str, const char prefix)
   return str.size() >= 1 && str[0] == prefix;
 }
 
+/*!
+ * \brief Removes \a suffix from the end of \a str, if present
+ * \param [in] str string to be searched
+ * \param [in] suffix string to check for
+ * \return     String with (one copy of) \a suffix removed from end, if it was present  
+ */
+inline std::string removeSuffix(const std::string& str, const std::string& suffix)
```
[the remark this answers is not in the document]
Comment (kennyweiss) [reply src/axom/core/utilities/StringUtilities.hpp:72]: Thanks @joshessman-llnl 

I created https://github.com/LLNL/axom/issues/624 to track this

File: src/tools/data_collection_util.cpp (hunk 11 above)
Comment (kennyweiss) [reply src/tools/data_collection_util.cpp:149]: Thanks for the suggestion.
I considered it, but I think it would be overkill in this case.
There are actually two different modes, and each one only has 3-ish main options.
e.g.
```bash
> srun -n8 ./bin/data_collection_util -f file -m star-q2.mesh --scale 5
> srun -n8 ./bin/data_collection_util --min 0 0 --max 1 1 --res 64 64 -p 2
```

I split these up into cli's `option_group`s, which provides a better help/usage message:
```bash
>./bin/data_collection_util -h
Utility tool to create a blueprint compliant data store
Usage: ./bin/data_collection_util [OPTIONS]

Options:
  -h,--help                                       Print this help message and exit
  -o,--output-name TEXT                           Name of the output mesh. Defaults to box{2,3}d for box meshes, and the mfem mesh name when loading from an mfem file
  -v,--verbose                                    Enable/disable verbose output
  -l,--ref-level INT:NONNEGATIVE=0                The number of uniform refinement levels to apply to the mesh
  -f,--mesh-form ENUM:value in {box->0,file->1} OR {0,1}=0
                                                  Determines the input type -- either box or file
[Option Group: box]
  Options for setting up a box mesh
  Options:
    --min FLOAT x 2 Needs: --max                    Min bounds for box mesh (x,y[,z])
    --max FLOAT x 2 Needs: --min                    Max bounds for box mesh (x,y[,z])
    --res INT x 2                                   Resolution of the box mesh (i,j[,k])
    -d,--dimension INT:POSITIVE                     Dimension of the box mesh
    -p,--polynomial-order INT:NONNEGATIVE=2         polynomial order of the generated mesh
[Option Group: file]
  Options for loading from an mfem mesh file
  Options:
    -m,--mfem-file TEXT:FILE                        Path to a mesh file in the mfem format
    --scale FLOAT                                   Scale factors for the file mesh. Can either be one value or mesh.Dimension() values
    --translate FLOAT x 2                           Translation for the file mesh (sx,sy[,sz])
```

File: src/axom/core/utilities/StringUtilities.cpp (hunk 1 above)
Comment (kennyweiss) [reply src/axom/core/utilities/StringUtilities.cpp:51]: See #624

**Review by white238 — APPROVED** [review]
